File: api/src/main/java/mg/fizanakara/api/services/TributeService.java
package mg.fizanakara.api.services;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import mg.fizanakara.api.dto.tributes.TributeDto;
import mg.fizanakara.api.models.Tribute;
import mg.fizanakara.api.repository.TributeRepository;
import org.springframework.data.rest.webmvc.ResourceNotFoundException;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class TributeService {
    private final TributeRepository tributeRepository;

    // GET ALL
    public List<Tribute> getAllTributes(){
        return tributeRepository.findAll();
    }

    // GET BY ID
    public Tribute getTributeByID(Long id){
        return tributeRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Tribute not found"));
    }

    // CREATE
    @Transactional
    public Tribute createTribute(TributeDto dto){
        if(tributeRepository.existsByName(dto.getName())){
            throw new IllegalArgumentException("Tribute of name " + dto.getName() + "is exist");
        }
        log.info("Creating Tribute : {}", dto.getName());
        Tribute tribute = Tribute.builder()
                .name(dto.getName())
                .build();
        return tributeRepository.save(tribute);
    }

    // UPDATE
    @Transactional
    public Tribute updateTribute(Long id, TributeDto dto){
        Tribute tribute = getTributeByID(id);
        if(!tribute.getName().equals(dto.getName()) && tributeRepository.existsByName(dto.getName())){
            throw new IllegalArgumentException("Tribute of name " + dto.getName() + " is exist");
        }
        log.info("Updating Tribute ID {} : {}", id, dto.getName());
        tribute.setName(dto.getName());
        return tributeRepository.save(tribute);
    }

    // DELETE
    @Transactional
    public void deleteTribute(Long id){
        Tribute tribute = getTributeByID(id);
        tributeRepository.delete(tribute);
    }

    // DELETE ALL
    public ResponseEntity<Map<String, Object>> deleteAllTributes() {
        tributeRepository.deleteAll();
        return ResponseEntity.ok(Map.of("message", "All districts deleted", "success", true));
    }
}

File: client/src/services/tribute.service.ts
import api from '../api/axios.config';
import { TributeDto } from '../lib/types/models/common.type';

const BASE_URL = '/api/admins/tributes';

export const TributeService = {
    getAll: async (): Promise<any[]> => {
        const response = await api.get(BASE_URL);
        return response.data;
    },

    create: async (data: TributeDto): Promise<any> => {
        const response = await api.post(BASE_URL, data);
        return response.data;
    },

    update: async (id: number, data: TributeDto): Promise<any> => {
        const response = await api.put(`${BASE_URL}/${id}`, data);
        return response.data;
    },

    delete: async (id: number): Promise<void> => {
        await api.delete(`${BASE_URL}/${id}`);
    }
};